extras/gizmo/axis-shapes.js

```diff
@@ -0,0 +1,650 @@
+
+// constants
+const SHADOW_DAMP_SCALE = 0.25;
+const SHADOW_DAMP_OFFSET = 0.75;
+const TORUS_RENDER_SEGMENTS = 80;
+const TORUS_INTERSECT_SEGMENTS = 20;
+const LIGHT_DIR = new Vec3(1, 2, 3);
+const MESH_TEMPLATES = {
+    box: createBox,
+    cone: createCone,
+    cylinder: createCylinder,
+    plane: createPlane,
+    torus: createTorus
+};
+
+// temporary variables
+const tmpV1 = new Vec3();
+const tmpV2 = new Vec3();
+const tmpQ1 = new Quat();
+
+function createTorus(device, opts = {}) {
+    // Check the supplied options and provide defaults for unspecified ones
+    const rc = opts.tubeRadius ?? 0.2;
+    const rt = opts.ringRadius ?? 0.3;
+    const segments = opts.segments ?? 30;
+    const sides = opts.sides ?? 20;
+    const sectorAngle = opts.sectorAngle ?? 2 * Math.PI;
+
+    // Variable declarations
+    const positions = [];
+    const normals = [];
+    const uvs = [];
+    const indices = [];
+
+    for (let i = 0; i <= sides; i++) {
+        for (let j = 0; j <= segments; j++) {
+            const x = Math.cos(sectorAngle * j / segments) * (rt + rc * Math.cos(2 * Math.PI * i / sides));
+            const y = Math.sin(2 * Math.PI * i / sides) * rc;
+            const z = Math.sin(sectorAngle * j / segments) * (rt + rc * Math.cos(2 * Math.PI * i / sides));
+
+            const nx = Math.cos(sectorAngle * j / segments) * Math.cos(2 * Math.PI * i / sides);
+            const ny = Math.sin(2 * Math.PI * i / sides);
+            const nz = Math.sin(sectorAngle * j / segments) * Math.cos(2 * Math.PI * i / sides);
+
+            const u = i / sides;
+            const v = 1 - j / segments;
+
+            positions.push(x, y, z);
+            normals.push(nx, ny, nz);
+            uvs.push(u, 1.0 - v);
+
+            if ((i < sides) && (j < segments)) {
+                const first  = ((i))     * (segments + 1) + ((j));
+                const second = ((i + 1)) * (segments + 1) + ((j));
+                const third  = ((i))     * (segments + 1) + ((j + 1));
+                const fourth = ((i + 1)) * (segments + 1) + ((j + 1));
+
+                indices.push(first, second, third);
+                indices.push(second, fourth, third);
+            }
+        }
+    }
+
+    const options = {
+        normals: normals,
+        uvs: uvs,
+        uvs1: uvs,
+        indices: indices
+    };
+
+    return createMesh(device, positions, options);
+}
+
+function createShadowMesh(device, entity, type, templateOpts = {}) {
+    const createTemplate = MESH_TEMPLATES[type];
+    if (!createTemplate) {
+        throw new Error('Invalid primitive type.');
+    }
+
+    const mesh = createTemplate(device, templateOpts);
+    const options = {
+        positions: [],
+        normals: [],
+        uvs: [],
+        indices: [],
+        colors: []
+    };
+
+    mesh.getPositions(options.positions);
+    mesh.getNormals(options.normals);
+    mesh.getIndices(options.indices);
+    mesh.getUvs(0, options.uvs);
```

Comment:
the gizmo mesh is not textured, does it need UVs? How about normals .. does it need those?

Reply:
The UVs are not needed but the normals are required for shadow calculation